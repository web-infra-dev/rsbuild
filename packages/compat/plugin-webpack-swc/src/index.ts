export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
