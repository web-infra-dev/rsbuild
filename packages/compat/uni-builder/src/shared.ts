import { deepmerge } from '@rsbuild/shared/deepmerge';
import type {
  RsbuildPlugin,
  RsbuildConfig as RsbuildRspackConfig,
} from '@rsbuild/core';
import type { RsbuildConfig as RsbuildWebpackConfig } from '@rsbuild/webpack';
import type { UniBuilderRspackConfig, UniBuilderWebpackConfig } from './types';

export function parseCommonConfig<B = 'rspack' | 'webpack'>(
  uniBuilderConfig: B extends 'rspack'
    ? UniBuilderRspackConfig
    : UniBuilderWebpackConfig,
): {
  rsbuildConfig: B extends 'rspack'
    ? RsbuildRspackConfig
    : RsbuildWebpackConfig;
  rsbuildPlugins: RsbuildPlugin[];
} {
  const rsbuildConfig = deepmerge({}, uniBuilderConfig);
  const { output } = rsbuildConfig;

  if (output.cssModuleLocalIdentName) {
    output.cssModules ||= {};
    output.cssModules.localIdentName = output.cssModuleLocalIdentName;
    delete output.cssModuleLocalIdentName;
  }

  return {
    rsbuildConfig: rsbuildConfig,
    rsbuildPlugins: [],
  };
}
