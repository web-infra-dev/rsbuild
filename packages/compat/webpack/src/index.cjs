/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
