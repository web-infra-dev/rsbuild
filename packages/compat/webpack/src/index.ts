export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
