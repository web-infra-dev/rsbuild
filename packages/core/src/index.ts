/**
 * The methods and types exported from this file are considered as
 * the public API of @rsbuild/core.
 */
import { rspack } from '@rspack/core';
import type * as Rspack from '@rspack/core';
import * as __internalHelper from './internal';

// Core methods
export { loadEnv } from './loadEnv';
export { createRsbuild } from './createRsbuild';
export { loadConfig, defineConfig } from './config';

// Rsbuild version
export const version = RSBUILD_VERSION;

// Rspack instance
export { rspack };
export type { Rspack };

// Helpers
export { logger } from '@rsbuild/shared';
export { mergeRsbuildConfig } from './mergeConfig';

// Constants
export { PLUGIN_SWC_NAME, PLUGIN_CSS_NAME } from './constants';

// Types
export type {
  // Config Types
  RsbuildConfig,
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  SourceConfig,
  ServerConfig,
  OutputConfig,
  SecurityConfig,
  PerformanceConfig,
  ModuleFederationConfig,
  // Normalized Config Types
  NormalizedConfig,
  NormalizedDevConfig,
  NormalizedHtmlConfig,
  NormalizedToolsConfig,
  NormalizedSourceConfig,
  NormalizedServerConfig,
  NormalizedOutputConfig,
  NormalizedSecurityConfig,
  NormalizedPerformanceConfig,
  NormalizedModuleFederationConfig,
  // Plugin Types
  RsbuildPlugin,
  RsbuildPlugins,
  RsbuildPluginAPI,
} from './types';

export type {
  RsbuildMode,
  RsbuildEntry,
  RsbuildTarget,
  RsbuildContext,
  RsbuildInstance,
  CreateRsbuildOptions,
  InspectConfigOptions,
  // Subtypes of Config
  Minify,
  Polyfill,
  PrintUrls,
  PublicDir,
  Decorators,
  RspackRule,
  WatchFiles,
  CSSModules,
  CrossOrigin,
  ConsoleType,
  SplitChunks,
  BundlerChain,
  ClientConfig,
  ScriptInject,
  ChainedConfig,
  PostCSSPlugin,
  ScriptLoading,
  LegalComments,
  AliasStrategy,
  FilenameConfig,
  DistPathConfig,
  OutputStructure,
  ChainIdentifier,
  PreconnectOption,
  CSSLoaderOptions,
  ModifyChainUtils,
  StyleLoaderOptions,
  PostCSSLoaderOptions,
  ChainedConfigWithUtils,
  ModifyRspackConfigUtils,
  CSSModulesLocalsConvention,
  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnAfterStartProdServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeStartProdServerFn,
  OnBeforeCreateCompilerFn,
  OnCloseDevServerFn,
  OnDevCompileDoneFn,
  ModifyBundlerChainFn,
  ModifyRspackConfigFn,
  ModifyRsbuildConfigFn,
  TransformFn,
  TransformHandler,
} from '@rsbuild/shared';

export {
  /**
   * @private
   */
  __internalHelper,
};
