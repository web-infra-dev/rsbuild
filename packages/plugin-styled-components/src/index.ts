import {
  isServerTarget,
  mergeChainedOptions,
  getDefaultStyledComponentsConfig,
  type ChainedConfig,
  type DefaultRsbuildPlugin,
} from '@rsbuild/shared';

/**
 * the options of [babel-plugin-styled-components](https://github.com/styled-components/babel-plugin-styled-components) or [rspackExperiments.styledComponents](https://rspack.dev/guide/loader.html#optionsrspackexperimentsstyledcomponents).
 */
type StyledComponentsOptions = {
  displayName?: boolean;
  ssr?: boolean;
  fileName?: boolean;
  meaninglessFileNames?: string[];
  namespace?: string;
  topLevelImportPaths?: string[];
  transpileTemplateLiterals?: boolean;
  minify?: boolean;
  pure?: boolean;
  cssProps?: boolean;
};

export const pluginStyledComponents = (
  userConfig: ChainedConfig<StyledComponentsOptions> = {},
): DefaultRsbuildPlugin => ({
  name: 'plugin-styled-components',

  setup(api) {
    api.modifyBundlerChain(async (chain, { CHAIN_ID, isProd }) => {
      const { bundlerType } = api.context;

      const isSSR = isServerTarget(api.context.target);

      const styledComponentsOptions = mergeChainedOptions({
        defaults: getDefaultStyledComponentsConfig(isProd, isSSR),
        options: userConfig,
      });

      if (!styledComponentsOptions) {
        return;
      }

      [CHAIN_ID.RULE.JS, CHAIN_ID.RULE.JS_DATA_URI].forEach((ruleId) => {
        if (chain.module.rules.has(ruleId)) {
          const rule = chain.module.rule(ruleId);
          // apply swc
          if (rule.uses.has(CHAIN_ID.USE.SWC)) {
            // apply rspack builtin:swc-loader
            if (bundlerType === 'rspack') {
              rule.use(CHAIN_ID.USE.SWC).tap((options) => {
                options.rspackExperiments ??= {};
                options.rspackExperiments.styledComponents =
                  styledComponentsOptions;
                return options;
              });
            } else {
              // apply webpack swc-plugin
              rule.use(CHAIN_ID.USE.SWC).tap((swc) => {
                swc.extensions.styledComponents = styledComponentsOptions;
                return swc;
              });
            }
          } else if (rule.uses.has(CHAIN_ID.USE.BABEL)) {
            // apply babel
            rule.use(CHAIN_ID.USE.BABEL).tap((babelConfig) => {
              babelConfig.plugins ??= [];
              babelConfig.plugins.push([
                require.resolve('babel-plugin-styled-components'),
                styledComponentsOptions,
              ]);
              return babelConfig;
            });
          }
        }
      });
    });
  },
});
