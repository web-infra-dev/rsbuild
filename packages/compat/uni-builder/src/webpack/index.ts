import {
  createRsbuild,
  type RsbuildConfig,
  type RsbuildPlugin,
  type RsbuildInstance,
} from '@rsbuild/core';
import type { RsbuildTarget } from '@rsbuild/shared';
import type { UniBuilderWebpackConfig } from '../types';
import type { CreateWebpackBuilderOptions } from '../types';
import { parseCommonConfig } from '../shared/parseCommonConfig';
import { pluginModuleScopes } from './plugins/moduleScopes';
import { pluginStyledComponents } from './plugins/styledComponents';
import { pluginBabel } from './plugins/babel';
import { pluginReact } from './plugins/react';
import { withDefaultConfig } from './defaults';

export async function parseConfig(
  uniBuilderConfig: UniBuilderWebpackConfig,
  cwd: string,
  frameworkConfigPath?: string,
  target?: RsbuildTarget | RsbuildTarget[],
): Promise<{
  rsbuildConfig: RsbuildConfig;
  rsbuildPlugins: RsbuildPlugin[];
}> {
  const { rsbuildConfig, rsbuildPlugins } = await parseCommonConfig<'webpack'>(
    uniBuilderConfig,
    cwd,
    frameworkConfigPath,
    target,
  );

  rsbuildPlugins.push(
    pluginBabel({
      babelLoaderOptions: uniBuilderConfig.tools?.babel,
    }),
  );
  rsbuildPlugins.push(pluginReact());

  if (uniBuilderConfig.tools?.tsLoader) {
    const { pluginTsLoader } = await import('./plugins/tsLoader');
    rsbuildPlugins.push(
      pluginTsLoader(
        uniBuilderConfig.tools.tsLoader,
        uniBuilderConfig.tools.babel,
      ),
    );
  }

  if (uniBuilderConfig.output?.enableAssetManifest) {
    const { pluginManifest } = await import('./plugins/manifest');
    rsbuildPlugins.push(pluginManifest());
  }

  if (uniBuilderConfig.security?.sri) {
    const { pluginSRI } = await import('./plugins/sri');
    rsbuildPlugins.push(pluginSRI(uniBuilderConfig.security?.sri));
  }

  if (uniBuilderConfig.experiments?.lazyCompilation) {
    const { pluginLazyCompilation } = await import('./plugins/lazyCompilation');
    rsbuildPlugins.push(
      pluginLazyCompilation(uniBuilderConfig.experiments?.lazyCompilation),
    );
  }

  rsbuildPlugins.push(
    pluginStyledComponents(uniBuilderConfig.tools?.styledComponents),
  );

  return {
    rsbuildConfig,
    rsbuildPlugins,
  };
}

export async function createWebpackBuilder(
  options: CreateWebpackBuilderOptions,
): Promise<RsbuildInstance> {
  const { rsbuildConfig, rsbuildPlugins } = await parseConfig(
    withDefaultConfig(options.config),
    options.cwd,
    options.frameworkConfigPath,
    options.target,
  );

  const { webpackProvider } = await import('@rsbuild/webpack');
  const rsbuild = await createRsbuild({
    rsbuildConfig,
    provider: webpackProvider,
    target: options.target || 'web',
  });

  rsbuild.addPlugins([
    ...rsbuildPlugins,
    pluginModuleScopes(options.config.source?.moduleScopes),
  ]);

  return rsbuild;
}
