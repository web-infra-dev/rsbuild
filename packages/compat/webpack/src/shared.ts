import { __internalHelper } from '@rsbuild/core';

const {
  getChainUtils,
  initRsbuildConfig,
  createDevServer,
  formatStats,
  getDevMiddleware,
  getStatsOptions,
  stringifyConfig,
  outputInspectConfigFiles,
} = __internalHelper;

export {
  getChainUtils,
  initRsbuildConfig,
  createDevServer,
  formatStats,
  getDevMiddleware,
  getStatsOptions,
  stringifyConfig,
  outputInspectConfigFiles,
};

export type InternalContext = __internalHelper.InternalContext;
