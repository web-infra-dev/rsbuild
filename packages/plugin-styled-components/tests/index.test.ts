import { expect, describe, it } from 'vitest';
import { pluginStyledComponents } from '../src';
import { createStubRsbuild } from '@rsbuild/test-helper';
import { SCRIPT_REGEX } from '@rsbuild/shared';
import { webpackProvider } from '@rsbuild/webpack';
import { pluginSwc } from '@rsbuild/plugin-swc';

describe('plugins/styled-components', () => {
  it('should enable ssr when target contain node', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
      target: ['node', 'web'],
    });

    rsbuild.addPlugins([pluginStyledComponents()]);
    const configs = await rsbuild.initConfigs();

    for (const config of configs) {
      expect(
        config.module.rules.find(
          (r) => r.test.toString() === SCRIPT_REGEX.toString(),
        ),
      ).toMatchSnapshot();
    }
  });

  it('should works in rspack mode', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
    });

    rsbuild.addPlugins([pluginStyledComponents()]);
    const config = await rsbuild.unwrapConfig();

    expect(
      config.module.rules.find(
        (r) => r.test.toString() === SCRIPT_REGEX.toString(),
      ),
    ).toMatchSnapshot();
  });

  it('should works in webpack babel mode', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
      provider: webpackProvider,
    });

    rsbuild.addPlugins([pluginStyledComponents()]);
    const config = await rsbuild.unwrapConfig();

    expect(
      config.module.rules.find(
        (r) => r.test && r.test.toString() === SCRIPT_REGEX.toString(),
      ),
    ).toMatchSnapshot();
  });

  // TODO move to @rsbuild/webpack
  it.skip('should works in webpack swc mode', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
      provider: webpackProvider,
    });

    rsbuild.addPlugins([pluginSwc(), pluginStyledComponents()]);
    const config = await rsbuild.unwrapConfig();

    expect(
      config.module.rules.find(
        (r) => r.test && r.test.toString() === SCRIPT_REGEX.toString(),
      ),
    ).toMatchSnapshot();
  });
});
