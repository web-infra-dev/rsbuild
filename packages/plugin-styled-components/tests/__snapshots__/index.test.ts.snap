// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/styled-components > should apply styledComponents option to swc-loader 1`] = `
{
  "include": [
    {
      "and": [
        "<ROOT>",
        {
          "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      ],
    },
    /\\\\\\.\\(\\?:ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
  ],
  "test": /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
  "type": "javascript/auto",
  "use": [
    {
      "loader": "builtin:swc-loader",
      "options": {
        "env": {
          "coreJs": "3.36",
          "mode": "usage",
          "shippedProposals": true,
          "targets": [
            "chrome >= 87",
            "edge >= 88",
            "firefox >= 78",
            "safari >= 14",
          ],
        },
        "isModule": "unknown",
        "jsc": {
          "externalHelpers": true,
          "parser": {
            "decorators": true,
            "syntax": "typescript",
            "tsx": false,
          },
          "preserveAllComments": true,
          "transform": {
            "decoratorMetadata": true,
            "legacyDecorator": true,
            "useDefineForClassFields": false,
          },
        },
        "rspackExperiments": {
          "styledComponents": {
            "displayName": true,
            "pure": false,
            "ssr": false,
            "transpileTemplateLiterals": true,
          },
        },
      },
    },
  ],
}
`;

exports[`plugins/styled-components > should enable ssr option when target contains node 1`] = `
{
  "include": [
    {
      "and": [
        "<ROOT>",
        {
          "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      ],
    },
    /\\\\\\.\\(\\?:ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
  ],
  "test": /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
  "type": "javascript/auto",
  "use": [
    {
      "loader": "builtin:swc-loader",
      "options": {
        "env": {
          "targets": [
            "node >= 16",
          ],
        },
        "isModule": "unknown",
        "jsc": {
          "externalHelpers": true,
          "parser": {
            "decorators": true,
            "syntax": "typescript",
            "tsx": false,
          },
          "preserveAllComments": true,
          "transform": {
            "decoratorMetadata": true,
            "legacyDecorator": true,
            "useDefineForClassFields": false,
          },
        },
        "rspackExperiments": {
          "styledComponents": {
            "displayName": true,
            "pure": false,
            "ssr": true,
            "transpileTemplateLiterals": true,
          },
        },
      },
    },
  ],
}
`;

exports[`plugins/styled-components > should enable ssr option when target contains node 2`] = `
{
  "include": [
    {
      "and": [
        "<ROOT>",
        {
          "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      ],
    },
    /\\\\\\.\\(\\?:ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
  ],
  "test": /\\\\\\.\\(\\?:js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
  "type": "javascript/auto",
  "use": [
    {
      "loader": "builtin:swc-loader",
      "options": {
        "env": {
          "coreJs": "3.36",
          "mode": "usage",
          "shippedProposals": true,
          "targets": [
            "chrome >= 87",
            "edge >= 88",
            "firefox >= 78",
            "safari >= 14",
          ],
        },
        "isModule": "unknown",
        "jsc": {
          "externalHelpers": true,
          "parser": {
            "decorators": true,
            "syntax": "typescript",
            "tsx": false,
          },
          "preserveAllComments": true,
          "transform": {
            "decoratorMetadata": true,
            "legacyDecorator": true,
            "useDefineForClassFields": false,
          },
        },
        "rspackExperiments": {
          "styledComponents": {
            "displayName": true,
            "pure": false,
            "ssr": true,
            "transpileTemplateLiterals": true,
          },
        },
      },
    },
  ],
}
`;
